### ISSUE TITLE
The mouse rotation in the gui when the screen is rotated is still not working

### ISSUE THREAD

**Issue by quincerly** [BlitterStudio/amiberry#1521]
Using the screen rotation option works (I use 90) but the mouse controls are still not rotated using the V7.0.0-RC1 Bookworm or Bullseye armhf debs on a Pi 3 with Bookworm/Bullseye PiOS 32bit installed via the rpi-imager.

I installed as above, set the screen rotation to 90 in .config/amiberry/amiberry.conf and then ran amiberry. Screen is rotated as expected but the mouse controls are not rotated in the GUI.

Love Amiberry? Please consider supporting it:
👉  https://ko-fi.com/midwan (I did) ) - Dan :o)

**Comment by midwan** [comment]
Are the controls rotated as expected inside emulation, however?
I think maybe I missed doing the same in the GUI.

**Comment by quincerly** [comment]
It looks to me that mouse controls are not rotated in the emulation either but I'm struggling as when I do rotate the display the scaling also goes wrong and the controlling the ui via keyboard to tweak setting does not seem to work predictably. Sorry this is not so helpful - I ran out of time this evening but will try to systematically test with a few photos and report here over the weekend.

**Comment by quincerly** [comment]
I can confirm that the screen is successfully rotated in both GUI and emulation with rotation_angle=90 but the mouse controls are not rotated in the GUI or the emulation.

I am also unable to get any usable screen scaling which I was able to get with Amiberry 3.0 (that is what I have been using until now). Could it be that more recent versions are unable to downscale the display to my 800x480 pixel screen (i.e. height<512)? The best I can get is with the LowRes option where the screen is cropped. I'm initially just trying to start a PAL HiRes (640x256) resolution Workbench screen. It worked  and looked amazing with version 3.0 but that does not support the rotation which I now need with my new display (Pimoroni Hyperpixel 4) and more recent versions of PiOS where other rotation options at the OS level are not working.

I am using it with X, i.e. booting my Pi to console without starting the desktop, and then running Amiberry from there.

**Comment by midwan** [comment]
Haven't had a chance to test this, but just as an idea:
Have you tried setting up the rotation on the OS/system level, instead of Amiberry? That might work better.
I believe the RPI `config.txt` itself had some rotation options, which got applied from boot.

**Comment by quincerly** [comment]
Thanks. It's a good idea, and its exactly how I managed to get older versions working. I can still get V3 of Amiberry working this way with ancient PiOS (Buster I think) and I'm stuck with that for now. Despite hours fiddling around and searching online for the solution I've not been able to get this approach to work with more recent, supported, versions of PiOS and the Hyperpixel screen, hence my excitement at the option in more recent versions of Amiberry. It will rotate the linux console screen but not Amiberry running from the console.

**Comment by midwan** [comment]
I can't really recreate the same environment here, but here's one idea:
- The old v3 version you mentioned that works, did not use any rotation information when opening the GUI/emulation screen.
- I could prepare an experimental build of v7, with the rotation options removed, to see if that helps.
- I assume you're using the latest Bookworm edition for all this, and you have already rotated the display in the OS itself?

Let me know if the above makes sense, and I will prepare an experimental binary for you (Bookworm 32-bit ARMHF, RPI3, from what I read above - please correct me if I'm wrong).

**Comment by midwan** [comment]
@quincerly 
Any feedback on the above?

**Comment by quincerly** [comment]
Thanks, sorry, not had a chance to follow up due to work and Christmas rush. Planning to find more time to give this over the Christmas break and will get back to you. I appreciate the support.

**Comment by quincerly** [comment]
>    The old v3 version you mentioned that works, did not use any rotation information when opening the GUI/emulation screen.
Yes, that works when the OS rotation works and so far I have only been able to get that working with very old versions of the OS (Buster). On newer versions of Raspberry Pi OS the screen rotation works in the text console but when Amiberry starts the screen is not rotated. This is why I when trying to get things working on more recent Raspberry Pi OS versions I was keen try Amiberry's roation option after many hours failing to get the OS level screen rotation working in OS versions more recent that Buster.

>    I could prepare an experimental build of v7, with the rotation options removed, to see if that helps.
Thanks, but for the above reason I don't that would help as I can't rotate the screen at the OS level on current Raspberry Pi OS versions.

>    I assume you're using the latest Bookworm edition for all this, and you have already rotated the display in the OS itself?
As mentioned above, I have only had success with OS level rotation with Buster and older. 

I'd be happy to get it working either with OS rotation (out of your expertise I realised and need the specific Hyperpixel GPIO display to resolve what's not workin), or with Amiberry's option, but I've run out of ideas at present.

> Let me know if the above makes sense, and I will prepare an experimental binary for you (Bookworm 32-bit ARMHF, RPI3, from what I read above - please correct me if I'm wrong).
As explained, I don't think this would be worth your effort unless I manage to get OS level rotation working. It seems the way the drivers for the Hyperpixel display work with screen rotation has changed since Buster and with the newer version of the Hyperpixel display.

So my current working setup is Raspbian Buster with Amiberry V3. It performs very well but clearly it would be good to move to supported OS versions and current Amiberry too. Your screen rotation does seem to be working on my setup with the latest Raspbian and Amiberry, but the mouse control rotation does not seem to be working.

**Comment by midwan** [comment]
@quincerly 
Are you using KMSDRM when running Amiberry? That might account for the differences, as I'm usually testing things under Wayland here.

**Comment by midwan** [comment]
From what I read so far, it seems it's not possible to rotate the screen under the OS itself, in newer versions. You can do it under X/Wayland, which is where the Hyperpixel docs on the newer v4 mention you should do it, also: https://shop.pimoroni.com/products/hyperpixel-4?variant=12569485443155

Under KMSDRM, the mouse in the GUI is controlled by the OS, so even if I rotate the Amiberry GUI, mouse input will not be rotated. It should still work with keyboard/joystick control however. Inside the emulation itself, we use relative mouse motion events, so those can also be rotated (unless you use the Virtual mouse option).

So, to summarize, it seems you can either use:
- KMSDRM, with Amiberry rotation options - the GUI will still not have the mouse rotated, but inside emulation it _should_ move as expected
- X/Wayland with the system rotation option, Amiberry set to _NOT_ rotate. Should work as expected everywhere.

**Comment by quincerly** [comment]
Thanks. I did try with X at some point but didn't have kmuch luck, especially when trying with a minimal window manager (matchbox) to avoid having to boot into a desktop first and to run full screen. I will revisit it though when I get a chance and report back how it goes.
